test/schema_test.exs

```diff
@@ -138,4 +139,60 @@ defmodule OpenApiSpex.SchemaTest do
     assert :ok = Schema.validate(schema, "bar", %{})
   end
 
+  test "Validate schema type object when value is array" do
+    schema = %Schema{
+      type: :object
+    }
+    assert {:error, _} = Schema.validate(schema, [], %{})
+  end
+
+  test "Validate schema type array when value is object" do
+    schema = %Schema{
+      type: :array
+    }
+    assert {:error, _} = Schema.validate(schema, %{}, %{})
+  end
+
+  test "Validate schema type boolean when value is object" do
+    schema = %Schema{
+      type: :boolean
+    }
+    assert {:error, _} = Schema.validate(schema, %{}, %{})
+  end
+
+  test "Validate schema type string when value is object" do
+    schema = %Schema{
+      type: :string
+    }
+    assert {:error, _} = Schema.validate(schema, %{}, %{})
+  end
+
+  test "Validate schema type string when value is DateTime" do
+    schema = %Schema{
+      type: :string
```

Comment:
mbuhot 6 hours ago  Collaborator
For this to be valid, I think it would require a format: :"date-time" in the schema.

  @tapickell
tapickell an hour ago
I had a question about that.
When I made the update to the validation on string type it broke other tests that made assertions on a DateTime being handled by the string type validation.

  1) test User Schema example matches schema (OpenApiSpex.SchemaTest)
     test/schema_test.exs:58
     Value does not conform to schema User: #/inserted_at: invalid type object where string expected
     %{"email" => "joe@gmail.com", "id" => 123, "inserted_at" => "2017-09-12T12:34:55Z", "name" => "Joe User", "updated_at" => "2017-09-13T10:11:12Z"}
     code: assert_schema(Schemas.User.schema().example, "User", spec)
     stacktrace:
       (open_api_spex) lib/open_api_spex/test/assertions.ex:32: OpenApiSpex.Test.Assertions.assert_schema/3
       test/schema_test.exs:60: (test)
This test would generate an error like this
{:error, "#/inserted_at: invalid type object where string expected"}

So for the moment I added testing to assert/document the validation of DateTime via
string type validator.
I would think DateTime should warrant it's own type validation.
Should the DateTime validations be handled just like a string?
Should there be a catch so a DateTime will error if it is validated as an object?

Reply:
> I would think DateTime should warrant it's own type validation.
> Should the DateTime validations be handled just like a string?
> Should there be a catch so a DateTime will error if it is validated as an object?

Yes, I suggest handling `DateTime` and `Date` as special cases of `type: :string` with `format: :"date-time"` and `format: :date`

Reply:
Something along the lines of:

```
def validate(schema = %Schema{type: :string}, value, path, _schemas) when is_binary(value) do
 ...
end
def validate(schema = %Schema{type: :string, format: :"date-time"}, value = %DateTime{}, path, _schemas) do
  :ok
end
def validate(schema = %Schema{}, value = %DateTime{}, path, _schemas) do
  {:error, "#{path}: DateTime can only be validated against type: :string, format: date-time"}
end
```

Reply:
:+1: Thanks @mbuhot I'll get that updated today.